Merge Tool and McpAIFunction to parse the complete InputSchema

Richasy opened the pull request:
# Pull Request

## Description of Changes

~~The definition of items is consistent with the current JsonSchema definition, just add it.~~

Update 2025/03/20:

According to the discussion in the comments, we may should use `JsonElement` instead of the predefined `JsonSchema` to avoid losing information during parsing. On this basis, we can merge the current `Tool` definition and `McpAIFunction` definition to simplify the code.

## Related Issue(s)
<!-- Please link to the issue(s) this PR addresses -->
- Fixes #104 

## Testing Done
<!-- Describe the testing you have performed -->
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [x] Manual testing performed

## Breaking Changes
<!-- List any breaking changes here, or delete this section if none -->
None

Richasy committed:
Add parsing of items in JsonSchemaProperty

stephentoub commented on src/mcpdotnet/Protocol/Types/JsonSchemaProperty.cs:
```diff
@@ -17,4 +17,10 @@ public class JsonSchemaProperty
+    /// Item type definition when type is <c>array</c>.
+    /// </summary>
+    [System.Text.Json.Serialization.JsonPropertyName("items")]
+    public JsonSchema? Items { get; set; }
```
@eiriktsarpalis, should this schema stuff be simplified to just have a JsonElement at a higher layer, so that anything provided as part of a tool's schema flows through?

eiriktsarpalis replied:
It depends on what the MCP specification itself allows. If it permits full-blown schemas then using `JsonElement` (with light validation) is fine.

stephentoub replied:
As far as I can tell, this is the only word on the subject:
- https://github.com/modelcontextprotocol/specification/blob/eb4abdf2bb91e0d5afd94510741eadd416982350/schema/2024-11-05/schema.ts#L699-L706
- https://github.com/modelcontextprotocol/specification/blob/eb4abdf2bb91e0d5afd94510741eadd416982350/schema/2024-11-05/schema.json#L1986-L2012

@PederHP, do you know?

PederHP replied:
Yes, what just typing a reply with the same reference. It is really weakly typed. The reference `everything` server even puts a `default` property on the properties, so I would say going strictly by the spec, it is server-defined what the properties are schema-wise. Could be anything.

PederHP replied:
There's an open issue related to the vagueness of this part of the spec: https://github.com/modelcontextprotocol/specification/issues/156

eiriktsarpalis replied:
Judging by [this response](https://github.com/modelcontextprotocol/specification/issues/156#issuecomment-2657870227) they kept it simple because they couldn't afford to embed the entire meta-schema. We should treat this as a JSON schema (with validation that the top-level object is indeed of `"type": "object"`).

Richasy committed:
Combine Tool and McpAIFunction to McpFunction

When we cannot predict the specific schema returned by the server, the existing predefined JsonSchema is likely to be invalid, so it is replaced with JsonElement. When InputSchema becomes JsonElement, the definition of Tool is basically the same as AIFunction, so the two can be further merged. After that, we no longer need to worry about the conversion between Tool and AIFunction.

Richasy committed:
Rename McpFunction to McpTool
